Homework 19

eolsen3 opened the pull request:
I keep getting the message : Unable to find element with locator By.cssSelector: input[type='email'].
So now Homework 18 won't run, and neither will Homework 19. I've tried using other selectors, to no avail. What's the issue?

TestPro-Support committed:
Update String url for LoginTests

eolsen3 committed:
Merge test 1

eolsen3 committed:
fix hopefully

eolsen3 committed:
Homework18

eolsen3 committed:
Homework 18

Homework 18

eolsen3 committed:
Homework 18

Homework 18

eolsen3 committed:
Homework18

Added some to xml file. Wasn't taught how to do it.

eolsen3 committed:
Homework 18 Successful Test

Updated LoginTests and xml file. Didn't add methods, because I believe the homework wanted everything in the class.

eolsen3 committed:
Update BaseTest

eolsen3 committed:
Homework 18 and Base Test updates

eolsen3 committed:
Homework19

rajniss21 commented on src/test/resources/Homework18.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
+
```
HomeWork18 missing "BaseUrl" parameter:
`<parameter name = "BaseUrl" value="https://qa.koel.app/"/>`

rajniss21 commented on src/test/java/Homework19.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+   public void openPlaylist() {
+       WebElement clickTestProPlaylist = driver.findElement(By.cssSelector("[href='#!/playlist/64251']"));
```
Your defined CSS selector was unable to find the WebElement.
 `WebElement clickTestProPlaylist = driver.findElement(By.cssSelector("section[id='playlists']>ul>li:nth-child(3)"));`

eolsen3 replied:
Can you explain to me how you got i:nth-child(3)"))?
I can see up to that point.
![Screenshot 2023-07-18 at 10 29 27 AM](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/137566317/708262ff-44c8-4ca3-8393-2d807f35b5f6)

rajniss21 commented:
Hello,
I have checked your homework 18 & 19 But you have got this error while compiling the code. Please review your code, fix it, and complete your homework.

- In HomeWork19 you got this assertion error. Please review it.
- HomeWork18 failed was of missing BaseURL.

Best wishes from the HW team (Rajnish, rajnish@testpro.io)!

![Screenshot_43](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/1f2bff1f-a9f3-4e25-8568-ef7c3e43bafa)
![Screenshot_42](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/80b46696-d7d1-49bf-bba9-44c4dd370617)

eolsen3 committed:
Update to Homework 18 and 19

eolsen3 commented:
I updated Homework 18, and Homework 19 I keep getting an error that the boolean expects to return true, but it's false. I can watch it go through the test successfully. It's not saying it can't find the locator, I don't know what the problem is.
I also tried div.success.show with no success.

Berichicko commented:
> I updated Homework 18, and Homework 19 I keep getting an error that the boolean expects to return true, but it's false. I can watch it go through the test successfully. It's not saying it can't find the locator, I don't know what the problem is. I also tried div.success.show with no success.

You got this problem, please correct it. 

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/d0c38dda-c2c9-4c19-b3c0-770e27c9c127)

**HINT**

 `WebElement successMessage = driver.findElement(By.cssSelector("div.success.show"));`
 
 
![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/9273b1c1-eb40-4ffb-b183-80a162854475)


![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/0eba6344-4ecc-43c5-a577-6401ee522a06)

eolsen3 committed:
Homework 19. Again.

Hi, I don't know how to get responses to my responses on Githug, so I'll try here. I don't know how you're able to get this to run, even with "div.success.show". It is not working for me. Also, how did you get div.success.show? I don't see how I could have figured out to use that locator. I also don't see how you got#playlists > ul > li:nth-child(3)". Can you explain those to me and also help me figure out why I'm getting assertion errors even when I do what you suggest I do?

Berichicko commented:
You got this problem, please correct it. 


![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/d8e8cce5-9ad3-454f-8d55-90b6afe3c647)

_I will try to help you how easy it is to find locators:_

**1) These steps describe the process of opening the developer console in the Google Chrome browser and making changes to the styles of a popover that appears when hovering over a specific element on a web page.**

For example

 `By.cssSelector("div.success.show")`

![image](https://user-images.githubusercontent.com/104761482/235407463-672c5728-57d6-4eb8-ae1b-945c83f12f5f.png)

![image](https://user-images.githubusercontent.com/104761482/235407647-4457ba8f-eba2-4c8f-8f57-bae764eeba43.png)

Browse to the desired page
1.	Open the dev console - F12 on Windows/Linux or option + ⌘ + J on macOS
2.	Select the Sources tab in chrome inspector
3.	In the web browser window, hover over the desired element to initiate the popover
4.	Hit F8 on Windows/Linux (or fn + F8 on macOS or  CTRL+\) while the popover is showing. If you have clicked anywhere on the actual page F8 will do nothing. Your last click needs to be somewhere in the inspector, like the sources tab
5.	Go to the Elements tab in inspector
6.	Find your popover (it will be nested in the trigger element's HTML)
7.	Have fun modifying the CSS
8.	https://www.howtogeek.com/663950/what-your-function-keys-do-in-chrome-devtools/



**2) You can use this extension for easy locator search.**

 Ranorex Selocity

Ranorex Selocity is a free extension for Chrome DevTools. It allows you to create unique selectors through a smart selector generator. Feedback is created, allowing you to locate your web page's elements in the most secure manner. https://www.ranorex.com/selocity/browser-extension/

For example 

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/d7f9b33c-8692-4fd8-b680-fb0da31ffd7b)

**3) Test**

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/1830ff25-7a32-4185-8c43-0fe7fca3b2aa)


```
    @Test
    public void deletePlaylist () throws InterruptedException {
        String deletedPlaylistMessage = "Deleted playlist \"" + "Test Pro Playlist.\"";

        provideEmail("erika.olsen@testpro.io");
        providePassword("te$t$tudent");
        clickSubmit();
        openPlaylist();
        clickDeletePlaylistBtn();
        Assert.assertTrue(getDeletedPlaylistMsg().contains(deletedPlaylistMessage));
    }
```
`String deletedPlaylistMessage = "Deleted playlist \"" + "Test Pro Playlist.\"";`

In Java, strings can be constructed by enclosing characters in double quotes (" "). The backslash (\) before the second double quote (") is an escape character used to include a double quote as part of the string. Therefore, the resulting string includes the text Deleted playlist followed by the name of the playlist, which in this case is Test Pro Playlist.. The backslashes help represent the quotes as part of the string without ending the string prematurely.

**The value of this string is:**

**Deleted playlist "Test Pro Playlist."**

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/a3584365-7403-4c22-a707-74d1324d5206)

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/0052d659-647d-453e-9eb7-a1e25ade8f2e)

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/f3e0a995-1813-4198-a9bc-3b155d781f04)

eolsen3 committed:
Final Homework 19

Thanks for the feedback. #1 Tip doesn't work. I've tried a mac and a windows, and "hovering over desired element to initiate the popover" doesn't work. I can hover for hours, no popup happens. #2, I've been using Ranorex with varying degrees of success for weeks, but it didn't help me this time. I found a colleague to help me. He said the assert that the teachers used in class was a weird way to go about it, and gave me a different assert that actually works.

Berichicko commented:
Hello,

I have checked your homework 19, and I wanted to let you know that you're doing great! Keep up the excellent work!
Best wishes from the HW team (Alex, [aliaksandr@testpro.io](mailto:aliaksandr@testpro.io))!

![804deb64dc406a20a75da05d91a1a69e](https://user-images.githubusercontent.com/104761482/235961479-9afb2841-b657-46c9-99f8-2402f933b843.jpg)

**Excellent! Perfect! Brilliant! Fantastic! Fantastic job! Magnificent! Terrific! Marvellous. Superb! Wow! Great stuff! You’re doing great! Keep it up! Very fine. Hooray for you! That’s better than ever! Super work!** 

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/fc00a84d-6da1-46da-ba4f-0c0a6247fae9)

eolsen3 committed:
Homework 19/20

During lesson 20 I had already started making changes to 19, so I renamed homework 19 branch and file to Homework 19_20 since I sort of messed that up, I guess.

eolsen3 committed:
Updates

eolsen3 committed:
Homework 19_20
